sklearn/semi_supervised/tests/test_self_training.py

```diff
@@ -0,0 +1,165 @@
+import numpy as np
+from sklearn.utils.testing import assert_array_equal
+from sklearn.utils.testing import assert_raise_message
+from sklearn.utils.testing import assert_warns_message
+from sklearn.exceptions import NotFittedError
+
+from sklearn.semi_supervised import SelfTrainingClassifier
+from sklearn.dummy import DummyClassifier
+from sklearn.model_selection import ParameterGrid
+from sklearn.neighbors import KNeighborsClassifier
+from sklearn.tree import DecisionTreeClassifier
+from sklearn.svm import SVC
+from sklearn.model_selection import train_test_split
+from sklearn.datasets import load_iris
+from sklearn.utils import check_random_state
+from sklearn.preprocessing import LabelBinarizer
+
+# Author: Oliver Rausch <rauscho@ethz.ch>
+# License: BSD 3 clause
+
+
+rng = check_random_state(0)
+
+# load the iris dataset and randomly permute it
+iris = load_iris()
+perm = rng.permutation(iris.target.size)
+iris.data = iris.data[perm]
+iris.target = iris.target[perm]
+X_train, X_test, y_train, y_test = train_test_split(iris.data,
+                                                    iris.target,
+                                                    random_state=rng)
+
+limit = 50
+
+y_train_missing_labels = y_train.copy()
+y_train_missing_labels[limit:] = -1
+y_train_missing_labels_strings = y_train_missing_labels.copy()
+mapping = {0: 'A', 1: 'B', 2: 'C', -1: -1}
+y_train_missing_strings = np.vectorize(mapping.get)(y_train_missing_labels)
+
+lb = LabelBinarizer()
+y_train_missing_dummies = lb.fit_transform(y_train_missing_labels)
+
+
+def test_classification():
```

Comment:
This test could benefit from using `@pytest.mark.parametrize`, see for example [here](https://github.com/scikit-learn/scikit-learn/blob/2d232acdeb66f1f983d389115de83c33e23b1276/sklearn/preprocessing/tests/test_data.py#L708)
 (don't bother about what the test is actually doing, this is just for the syntax)

That would avoid the 2 for loops and the use of ParameterGrid, because pytest would cross-product every parameter itself.

Same for other tests like `test_invalid_params`.